File: 0-99/19-remove-nth-node-from-end-of-list.py
# coding: utf-8

# Definition for singly-linked list.
# class ListNode:
#	 def __init__(self, x):
#		 self.val = x
#		 self.next = None

from made_list import ListNode

class Solution:
	# @return a ListNode
	def removeNthFromEnd(self, head, n):
		pre = pos = head
		if pos.next == None:
			return None

		while pos != None:
			temp = pos
			for x in range(n):
				temp = temp.next

			if temp == None:		#删除pos
				if pos == head:
					head = head.next
					del pos
				else:
					pre.next = pos.next
					del pos
				return head

			pre = pos
			pos = pos.next
		return head

a = Solution()

h = ListNode('1->2->3->4')

print a.removeNthFromEnd(h, 2)

File: special-1-to-200/19-remove-nth-node-from-end-of-list.go
package main

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	var (
		front, end = head, head
	)
	for i := 0; i < n; i++ {
		end = end.Next
	}
	// 特殊情况，需要移除第一个元素
	if end == nil {
		return front.Next
	}
	// front 和 end 相差 n
	// 当 end 在最后一个元素时，front 就指向要被移除元素的前一个元素
	for end.Next != nil {
		front, end = front.Next, end.Next
	}
	front.Next = front.Next.Next
	return head
}
